docs/features/incremental-generators.cookbook.md

```diff
@@ -636,7 +636,188 @@ TODO: https://github.com/dotnet/roslyn/issues/72149
+    {
+        context.RegisterPostInitializationOutput(ctx =>
+        {
+            //Generate the AutoImplementProperties Attribute
+            const string autoImplementAttributeDeclarationCode = $$"""
+// <auto-generated/>
+using System;
+namespace {{AttributeNameSpace}};
+
+[AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+sealed class {{AttributeClassName}} : Attribute
+{
+    public Type[] InterfacesTypes { get; }
+    public {{AttributeClassName}}(params Type[] interfacesTypes)
+    {
+        InterfacesTypes = interfacesTypes;
+    }
+}
+""";
+            ctx.AddSource($"{AttributeClassName}.g.cs", autoImplementAttributeDeclarationCode);
+        });
+
+        IncrementalValuesProvider<ClassModel> provider = context.SyntaxProvider.ForAttributeWithMetadataName(
+            fullyQualifiedMetadataName: FullyQualifiedAttributeName,
+            predicate: static (node, cancellationToken_) => node is ClassDeclarationSyntax,
+            transform: static (ctx, cancellationToken) =>
+            {
+                ISymbol classSymbol = ctx.SemanticModel.GetDeclaredSymbol(ctx.TargetNode, cancellationToken)!;
+
+                return new ClassModel(
+                    classSymbol.Name,
+                    classSymbol.ContainingNamespace.ToDisplayString(),
+                    GetInterfaceModels(classSymbol)
+                    );
+            }).Where(m => m is not null);
+
+        context.RegisterSourceOutput(provider, static (context, classModel) =>
+        {
+            foreach (InterfaceModel interfaceModel in classModel.Interfaces)
+            {
+                StringBuilder sourceBuilder = new($$"""
+                    // <auto-generated/>
+                    namespace {{classModel.NameSpace}};
+
+                    public partial class {{classModel.Name}} : {{interfaceModel.FullyQualifiedName}}
+                    {
+
+                    """);
+
+                foreach (string property in interfaceModel.Properties)
+                {
+                    sourceBuilder.AppendLine(property);
+                }
+
+                sourceBuilder.AppendLine("""
+                    }
+                    """);
+
+                //Concat class name and interface name to have unique file name if a class implements two interfaces with AutoImplement Attribute
+                string generatedFileName = $"{classModel.Name}_{interfaceModel.FullyQualifiedName}.g.cs";
+                context.AddSource(generatedFileName, sourceBuilder.ToString());
+            }
+        });
+    }
+
+    private static EquatableList<InterfaceModel> GetInterfaceModels(ISymbol classSymbol)
+    {
+        EquatableList<InterfaceModel> ret = [];
+
+        AttributeData attribute = classSymbol.GetAttributes().First(a => a.AttributeClass!.Name == AttributeClassName);
+        INamedTypeSymbol[] interfaceSymbols = attribute.ConstructorArguments[0].Values.Select(x => (x.Value as INamedTypeSymbol)!).ToArray();
```

Comment:
You can't just assume that everything bound correctly and is a non-null type. The user could have an error.

```suggestion
        // A separate analyzer should error on using a non-interface type in this list.
        INamedTypeSymbol[] interfaceSymbols = attribute.ConstructorArguments[0].Values.Select(x => x.Type).Where(t => t is { TypeKind: TypeKind.Interface }).ToArray()!;
```

Reply:
You are right about "You can't just assume that everything bound correctly and is a non-null type. The user could have an error.", but it seems your solution is not working.
1) Every element of "_**attribute.ConstructorArguments[0].Values**_" in this case has _**Type.TypeKind == TypeKind.Class**_
2) "_**attribute.ConstructorArguments[0].Values.Select(x => x.Type).Where(t => t is { TypeKind: TypeKind.Interface }).ToArray()!**_" is a "_**ITypeSymbol[]**_", not a "_**INamedTypeSymbol[]**_". So the subsequent foreach is not working, cause it finds properties whose type is assembly.

I'll work on this

Reply:
1. Well, then you're not passing an interface.
2. All interfaces are `INamedTypeSymbol`s, you can just add a `Cast<INamedTypeSymbol>()` and it'll work fine.

Reply:
Sorry, probably I'm not understanding something
![Screenshot 2024-06-21 213852](https://github.com/dotnet/roslyn/assets/92273914/b93ae2de-2eb8-44ac-bfee-df9fd7f3f626)
As you can see, even if I write "_**[AutoImplementProperties(typeof(IUserInterface), typeof(IUserInterface2))]**_" the TypeKind is "_**Class**_"

And both "_**tempo**_" and "**_tempo2_**" arrays have 61 elements (you can see some of them in the screenshot) instead of the 1 I expected, as every Interface in my sample only have 1 property

I'm keeping updated this repo in order to test every update I'm committing
https://github.com/bdbonazz/AutoImplementGenerator

Reply:
Ah, I think I see. The select should select the `Value`, not `Type`, and check for `INamedTypeSymbol` there. `Type` in this case is the actual type of the argument, which is `System.Type`, not the type bound by the `typeof`. `Value` has that.

Reply:
Perfect
I added the check on TypeKind.Interface